runners/samza/src/main/java/org/apache/beam/runners/samza/metrics/SamzaMetricsContainer.java

```diff
@@ -81,15 +106,133 @@ public MetricsContainerStepMap getContainers() {
+                    }
+                  },
+                  0,
+                  metricUpdateIntervalSec,
+                  TimeUnit.SECONDS);
+          executorServiceStarted = true;
+          LOG.info("Started executor service for periodic metrics updates.");
+        }
+      }
+    }
+  }
+
+  /**
+   * Shutdown the executor service and cancel the scheduled task. Before shutting down, update
+   * metrics one last time to ensure completeness.
+   */
+  private void shutdownExecutorService() {
+    if (executorService != null && !executorService.isShutdown()) {
+      LOG.info("Shutting down executor service...");
+
+      // Cancel the scheduled task if it's still running.
+      // This ensures that any periodic metrics updates are stopped and no further updates
+      // are scheduled once we begin the shutdown process.
+      if (scheduledFuture != null && !scheduledFuture.isCancelled()) {
+        LOG.info("Cancelling scheduled metrics updates...");
+        scheduledFuture.cancel(true);
+      }
+
+      // Update metrics one last time, ensuring that we're still committing from a single thread.
+      // This guarantees that all remaining metrics are committed before shutting down the executor
+      // service.
+      commitMetricsForAllSteps();
+
+      // Shutdown the executor service gracefully.
+      // Allow any currently executing tasks to finish, then terminate the service. If the shutdown
+      // process takes longer than 5 seconds, force a shutdown to ensure the service is stopped.
+      executorService.shutdown();
+      try {
+        if (!executorService.awaitTermination(5, TimeUnit.SECONDS)) {
+          LOG.warn("Forcing shutdown of executor service...");
+          executorService.shutdownNow();
+        }
+      } catch (InterruptedException e) {
+        LOG.error("Interrupted during shutdown, forcing shutdown now", e);
+        executorService.shutdownNow();
+        Thread.currentThread().interrupt();
+      }
+    }
+  }
 
-    assert metricsRegistry != null;
-    final List<String> stepNameList = Arrays.asList(stepName, GLOBAL_CONTAINER_STEP_NAME);
-    // Since global metrics do not belong to any step, we need to update it in every step.
-    final MetricResults metricResults =
+  /**
+   * This method commits metrics updates for all scheduled steps. It is invoked periodically by the
+   * scheduled executor service when `deferToExecutor` is set to true and also during shutdown to
+   * ensure that any remaining metrics are committed before terminating the executor.
+   */
+  private void commitMetricsForAllSteps() {
+    activeStepNames.forEach(this::updateMetricsInternal);
+  }
+
+  /**
+   * Private method containing the core logic for updating metrics. This is either called
+   * immediately or by the executor service, depending on the configuration.
+   *
+   * @param stepName the step name for which metrics are being updated
+   */
+  private void updateMetricsInternal(String stepName) {
```

Comment:
Is this method thread-safe?  Do we need thread-safety for this method?

Reply:
Good point. It’s possible that previously, only a single thread would enter updateMetricsInternal even with multiple instantiations. However, with the introduction of potential multiple thread pools (each committing metrics every second), we might indeed be introducing a race condition. For now, I’ll wrap the updateMetricsInternal call in a synchronized block during periodic evaluations to ensure thread safety and avoid potential race conditions.

That said, this is an experimental PR, and we can validate the code further after testing and understanding this codebase better. If we find that synchronization isn’t necessary.